modules/ca/src/tools/caget.c

```diff
@@ -283,73 +356,85 @@ static int caget (pv *pvs, int nPvs, RequestT request, OutputT format,
-                       dbr_type_to_text(pvs[n].dbrType));
-                if (pvs[n].dbrType == DBR_CLASS_NAME)
-                    printf("    Class Name:       %s\n",
-                           *((dbr_string_t*)dbr_value_ptr(pvs[n].value,
-                                                          pvs[n].dbrType)));
-                else {
-                    printf("    Element count:    %lu\n"
-                           "    Value:            ",
-                           pvs[n].nElems);
-                    if (charArrAsStr && dbr_type_is_CHAR(pvs[n].dbrType) && (reqElems || pvs[n].nElems > 1)) {
-                        dbr_char_t *s = (dbr_char_t*) dbr_value_ptr(pvs[n].value, pvs[n].dbrType);
-                        int dlen = epicsStrnEscapedFromRawSize((char*)s, strlen((char*)s));
-                        char *d = calloc(dlen+1, sizeof(char));
-                        if(d) {
-                            epicsStrnEscapedFromRaw(d, dlen+1, (char*)s, strlen((char*)s));
-                            printf("%s", d);
-                            free(d);
-                        } else {
-                            fprintf(stderr,"Failed to allocate space for escaped string\n");
-                        }
-                    } else {
-                        for (i=0; i<pvs[n].nElems; ++i) {
-                            if (i) printf ("%c", fieldSeparator);
-                            printf("%s", val2str(pvs[n].value, pvs[n].dbrType, i));
-                        }
-                    }
-                    printf("\n");
-                    if (pvs[n].dbrType > DBR_DOUBLE) /* Extended type extra info */
-                        printf("%s\n", dbr2str(pvs[n].value, pvs[n].dbrType));
-                }
-            }
-            break;
-        default :
-            break;
         }
+        break;
+    default :
+        break;
     }
+
+    epicsMutexUnlock(printMutex);
+
     return 0;
 }
 
+static void complainIfNotPlainAndSet (OutputT *current, const OutputT requested)
+{
+    if (*current != plain)
+        fprintf(stderr,
+                "Options t,d,a are mutually exclusive. "
+                "('caget -h' for help.)\n");
+    *current = requested;
+}
+/*
+void test(void* thread_data_void) {
+printf("called\n");
+    //t_thread_data* thread_data = (t_thread_data*)thread_data_void;
+    //epicsSpinLock(thread_data->spin);
+    //epicsMutexUnlock(thread_data->printMutex_lock_main);
+printf("End of test\n");
+}
+*/
+
+// Create a thread name based on the PV index in the order sent as argument for caget
+void thread_name (int pv_num, char* th_name) {
+    // Threads will be named caget0, caget1, ..., caget42, ...
+    char pv_num_str[3];
+    sprintf(pv_num_str, "%d", pv_num);
+    strcpy(th_name, "caget");
+    strcat(th_name, pv_num_str);
+}
+
+// Function that runs inside each thread, calling the caget function
+void caget_caller (void* thread_data_void)
+{
+    int result;                 /* CA result */
```

Comment:
There is a call at the end of the `main()` function below to `ca_context_destroy()` so I assume there is also one to `ca_context_create()` earlier on. I would like to see all of these caget threads sharing the same CA context, as that means that all channels opened to the same IOC will share the same TCP socket, which uses fewer resources on the IOC. To do that your code needs to get and save the newly-created CA context pointer by calling [`ca_current_context()`](https://epics.anl.gov/base/R7-0/8-docs/CAref.html#ca_current_context), and then each thread should call [`ca_attach_context()`](https://epics.anl.gov/base/R7-0/8-docs/CAref.html#ca_attach_context) to use it. When the thread is finished it should detach from the context again.

Reply:
The original caget code calls ca_context_create() indirectly via:
ca/src/tools/tool_lib.c
connect_pvs -> create_pvs ->

ca/src/client/access.cpp
ca_create_channel -> fetchClientContext -> epicsThreadPrivateGet -> ca_context_create

Also, ca_context_destroy() was in the original caget code. Maybe @ralphlange can help us figure out the reason faster than I, because right now I'm reverse engineering.

My changes made each thread call connect_pvs, which is creating one channel per thread. Maybe each thread should call ca_clear_channel() after printing the PV data and ca_context_destroy() continues in the end of the main function?

Reply:
Hi Márcio,

caget.c line 533 calls `ca_context_create(ca_disable_preemptive_callback)`, you were thinking of `ca_create_channel()`. The current version of the CA client library is multi-threaded internally, but the parameter passed to `ca_context_create()` controls whether the client application experiences the effects of that multi-threading or not; caget.c's setting `ca_disable_preemptive_callback` makes the API act like a single-threaded library in that the user callbacks will only be called while the main thread is executing inside the routines `ca_pend_io()`, `ca_poll()` or `ca_pend_event()`.

This mode has the advantage that you don't need to protect variables accessed in both the main thread and the callback routines with a mutex or atomic accesses. However a multi-threaded application can only share a context which was created using `ca_context_create(ca_enable_preemptive_callback)`. See the two sections of the CA reference manual starting [here](https://epics.anl.gov/base/R7-0/8-docs/CAref.html#Thread) for more information about the above.

I believe your code at the moment is creating a separate CA context for each thread. This is going to use significantly more resources, since each context creates multiple threads, internal data structures, and its own TCP and UDP sockets for doing name searches and connecting to IOCs. @mdavidsaver's raised eyebrow was related to this; I wasn't thinking about scaling to 10,000 channel names when I suggested that this approach might be valid.

You can still use the `ca_disable_preemptive_callback` mode to connect and fetch all the channel data if you re-work this. When using the callbacks instead of threads you don't actually need the mutexes that you added, the CA library takes care of the synchronization so that your code is never running multiple routines at the same time.

I hope you will be able to find some time to continue working on this.

Thanks,

- Andrew